src/dev/amdgpu/amdgpu_device.cc

```diff
@@ -846,16 +839,13 @@ AMDGPUDevice::unserialize(CheckpointIn &cp)
 
 
     if (doorbells_size > 0) {
-        uint32_t doorbells_offset[doorbells_size];
-        QueueType doorbells_queues[doorbells_size];
-        int doorbells_ip_ids[doorbells_size];
+        auto doorbells_offset = std::make_unique<uint32_t[]>(doorbells_size);
+        auto doorbells_queues = std::make_unique<QueueType[]>(doorbells_size);
+        auto doorbells_ip_ids = std::make_unique<int[]>(doorbells_size);
 
-        UNSERIALIZE_ARRAY(doorbells_offset, sizeof(doorbells_offset)/
-                sizeof(doorbells_offset[0]));
-        UNSERIALIZE_ARRAY(doorbells_queues, sizeof(doorbells_queues)/
-                sizeof(doorbells_queues[0]));
-        UNSERIALIZE_ARRAY(doorbells_ip_ids, sizeof(doorbells_ip_ids)/
-                sizeof(doorbells_ip_ids[0]));
+        UNSERIALIZE_UNIQUE_PTR_ARRAY(doorbells_offset, doorbells_size);
```

Comment:
I am not sure I've totally convinced myself why doorbells_size is guaranteed to be the same as doorbells_offset / sizeof(doorbells_offset[0]) in all cases ... am I missing something?

(this same question applies in other places too, but just marking here)

Reply:
In C/C++, `sizeof(a)/sizeof(a[0])` is an idiom to get the number of elements in an array when its size is known by the compiler. In fact it's often defined as a macro. `sizeof(a)` evaluates to the size of the whole array in bytes, and `sizeof(a[0])` evaluates to the size of the first element in bytes.

On lines 842..845 we've allocated 3 `doorbells_foo` arrays of size `doorbells_size`.